crates/telio-firewall/src/firewall.rs

```diff
@@ -608,16 +644,73 @@ impl StatefullFirewall {
     }
 
     fn handle_inbound_icmp<'a, P: IpPacket<'a>>(&self, peer: &PublicKey, ip: &P) -> bool {
-        let icmp_packet = unwrap_option_or_return!(P::Icmp::new(ip.payload()), false);
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), false);
+
+        let mut icmp_cache = unwrap_lock_or_return!(self.icmp.lock(), false);
+        let key = Self::build_icmp_key(ip, true);
+        if let Some(key) = key {
+            if icmp_cache.get(&key).is_some() {
+                telio_log_trace!("Matched ICMP conntrack entry {:?}", key,);
+                telio_log_trace!("Removing ICMP conntrack entry {:?}", key);
+                icmp_cache.remove(&key);
+                telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
+                return true;
+            }
+        }
 
-        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_type()) {
+        if P::Icmp::BLOCKED_TYPES.contains(&icmp_packet.get_icmp_type().0) {
             telio_log_trace!("Dropping ICMP packet {:?} {:?}", ip, peer);
             return false;
         }
 
         telio_log_trace!("Accepting ICMP packet {:?} {:?}", ip, peer);
         true
     }
+
+    fn build_icmp_key<'a, P: IpPacket<'a>>(ip: &P, inbound: bool) -> Option<IcmpConn> {
+        let icmp_packet = unwrap_option_or_return!(IcmpPacket::new(ip.payload()), None);
+        let (response_type, has_identifier_sequence) = match icmp_packet.get_icmp_type().0 {
+            // ICMP types
+            0 => (u8::MAX, true),  // echo reply message
+            8 => (0, true),        // echo request message
+            13 => (14, true),      // timestamp message
+            14 => (u8::MAX, true), // timestamp reply message
+            15 => (16, true),      // information request message
+            16 => (u8::MAX, true), // information reply message
+            17 => (18, true),      // address mask request message
+            18 => (u8::MAX, true), // address mask reply message
+            // ICMPv6 types
+            128 => (129, true),      // echo request message
+            129 => (u8::MAX, true),  // echo reply message
+            133 => (134, false),     // router solicit message
+            134 => (u8::MAX, false), // router advert message
+            135 => (136, false),     // neighbor solicit message
+            136 => (u8::MAX, false), // neighbor advert message
+            // No other types go into the cache
+            _ => return None,
+        };
```

Comment:
There is no test for this, if matching replies to requests work. You should add it. But I believe it does not work. Request and replies should produce the same key in the cache. Here outgoing echo request produces different key then incoming echo response, unless I'm missing something. I think it should be sth like:
```
            0 => (0, true),     // echo reply message
            8 => (0, true),     // echo request message
            13 => (14, true),   // timestamp message
            14 => (14, true),   // timestamp reply message
            ...
```
Additionally incoming requests should be just blocked without looking into the cache.

Reply:
But now thinking about it, we only have requests on `ICMP_BLOCKED_TYPES` lists, so actually the logic that we wanted is already implemented on main. The requests do not go through unless the peer is whitelisted and the responses do come through. What am I missing? :D

Reply:
You're right about the first comment here, I will update it and add a test.

Regarding the second comment here, the logic on main is: 
1. If whitelisted, allow
2. If on block list, reject
3. allow
and with the changes in this PR, the logic is:
1. if whitelisted, allow
2. if connection is a direct response, allow
3. if on block list, reject
4. allow

Reply:
TBH this seems strange to me :) Now the only added value I see is that we do not allow ICMP responses if the request was not initiated by this peer. But we allow all other ICMP packets in anyway, so I'm not sure what's so special about responses that we want to block them, but not block other icmp packets that are not req/resp. Also now the implementation contains a bug. I've hinted this in my first msg by: "Additionally incoming requests should be just blocked without looking into the cache." Because now the remote peer can send echo request to us when we send the request to him, and that for sure should be not allowed if the peer is not whitelisted.

Reply:
This should be fixed now. As before, all outbound packets are allowed, and for inbound packets, the logic is now:
1. If whitelisted, allow
2. If on block list, reject (the block list contains the request types we put in the cache, but not the reply types, so we block `EchoRequest` for example)
3. If direct response, allow
4. reject